Signal/src/AppDelegate.m

```diff
@@ -76,6 +76,9 @@ - (void)applicationDidEnterBackground:(UIApplication *)application {
 - (void)applicationWillEnterForeground:(UIApplication *)application {
     DDLogWarn(@"%@ applicationWillEnterForeground.", self.logTag);
 
+    // We want to prod OWSStorage here so that all storage is valid.
+    [OWSStorage applicationWillEnterForeground];
```

Comment:
Poking `OWSStorage` to re-open its database(s) should be the first thing we do when re-entering the background.  That's pretty easy to guarantee in the main app since the app delegate will be informed before anyone using `NSNotifications` to learn about this event.  In the SAE it's a little trickier since the SAE has to use notifications.  Therefore we have the app contexts fire lifecycle events.  This has two advantages:

* The SAE gets a chance to poke `OWSStorage` before any other component "awakens".
* We can unify main app and SAE lifecycle events by observing a single set of notifications.